File: 9.贪心算法/4.最大子序和/53.最大子数组和.cpp
/*
 * @lc app=leetcode.cn id=53 lang=cpp
 *
 * [53] 最大子数组和
 */
#include "tools.h"

// @lc code=start
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int n = nums.size();
        int maxS = nums[0], cur = nums[0];
        for (int i = 1; i < n; i ++ )
        {
            cur = max(nums[i], cur + nums[i]);
            maxS = max(cur, maxS);
        }
        return maxS;
    }
};
// @lc code=end

File: 9.贪心算法/4.最大子序和/53.最大子数组和.py
#
# @lc app=leetcode.cn id=53 lang=python3
#
# [53] 最大子数组和
#
from mytools import *
# @lc code=start
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        res = float('-inf')
        tmp = 0
        for i in range(len(nums)):
            tmp += nums[i]
            if tmp > res:
                res = tmp
            if tmp < 0: tmp = 0
        return res
# @lc code=end

